### ISSUE TITLE
Dynamic Secret

### ISSUE THREAD

**Issue by fgblomqvist** [vimalloc/flask-jwt-extended#121]
Something I came across the other day that I really like is per-user secrets for signing JWTs. It is (according to me) a great way to manage validation of dynamic JWTs.

Example:
You want to store a user's permissions in the JWT. This could be useful for the front-end so that it knows what to render for each user. The issue? Permissions can change frequently (and often server-side, such as when banning a user). The most automatic solution is to sign the JWT with a hash of `the secret + all of the dynamic data`. That way, the JWT will automatically fail to validate if the user's permissions change.

Another common use-case is if the user changes the password and you want to invalidate all the old JWTs. Solution? Sign the JWT with `the secret + the user password hash`.

Now, one could argue that this requires a DB query and that it defeats the purpose of JWTs, but geez that argument is getting old. It is not the sole purpose of JWTs to prevent DB queries.

What do you think? I could submit a PR for this (or at least try) if you just point the way in terms of naming. I think having a loader, e.g. `secret_loader`, is the smoothest solution here. It should take the verified/unverified claims as a parameter and return a string that the JWT should be validated/signed with (aka a secret). It would run right before encoding and decoding a token. Alternatively, you have a loader for the decoding, and then just add a secret param to the create_access_token function.

**Comment by vimalloc** [comment]
Are you familiar with the [claims_verification_loader](http://flask-jwt-extended.readthedocs.io/en/latest/api.html#flask_jwt_extended.JWTManager.claims_verification_loader) feature in this extension? I think you could accomplish the scenaiors outlined above using this. In the first case, you could check the permissions (or whatever other data) in the JWT against the database. For the second case, you could get a little cleaver and store the password revision (number of times the password has been changed) in the database and in the JWT, and compare those in this callback.

I'm not necessarily opposed to adding that functionality to this extension, but I would like to hear your thoughts on if it is actually necessary with the existing functionality this extension provides.

Cheers.

**Comment by fgblomqvist** [comment]
I am actually already using the claims_verification_loader to verify that it has the claims that I want. I suppose what you're suggesting would work, I would just extend it to verify that the data in the token is equal to the data in the DB. Since I'm not actually handling user passwords (Google OAuth) I don't care too much about solving the other problem (if I did, I probably would have wanted something different than counting how often it has changed, even though that is a clever solution).

Oh well, maybe someone will find this issue at some point and find whatever I wrote and you responded helpful! :)
